### DISCUSSION TITLE
Dynamically load wasm compiled assembly into dotnet

### DISCUSSION THREAD

**Discussion by bogdancucosel** [dotnet/runtime#95903] [Q&A] [answered]
Hi!

I am trying to understand if dynamically loading a wasm compiled assembly is possible in dotnet browser-wasm.
I tried something like:

```
const instance = await dotnet. ... .create();
let assemblyName = "wasmLibrary.wasm";
const response = await fetch("./" + assemblyName);
const assemblyBytes = await response.arrayBuffer();
const isPdb = instance.MONO.mono_wasm_add_assembly(assemblyName, assemblyBytes, assemblyBytes.byteLength);
```

But this fails, as the https://github.com/dotnet/runtime/blob/e9ddf98c8502cfa00823dc94d3fcbfe5aa03ece7/src/mono/mono/metadata/bundled-resources.c#L129 asserts for the root domain not be available. 

Is this something that could be possible ?

Thanks,
Bogdan

**Comment by maraf** [comment, marked as answer]
The `mono_wasm_add_assembly` can be used only before the runtime has started. You can leverage Blazor's lazy loading (outside of Blazor) with a bit of internal API. You can use `INTERNAL.loadLazyAssembly` to lazy load assembly marked as `<BlazorWebAssemblyLazyLoad Include="wasmLibrary.wasm"" />`.

**Reply by bogdancucosel** [reply]
Yes, I tried that, but that requires a prior knowledge of the assembly I need to load right ? 
I cannot modify the config after the runtime is loaded, correct ?
Thanks

**Reply by maraf** [reply]
If you want to do it dynamically, you can pass the downloaded byte array to C# and use `AssemblyLoadContext` https://github.com/dotnet/runtime/blob/main/src%2Flibraries%2FSystem.Runtime.InteropServices.JavaScript%2Fsrc%2FSystem%2FRuntime%2FInteropServices%2FJavaScript%2FJSHostImplementation.cs#L351-L357. Note that you will have to solve deployment on your own

**Reply by bogdancucosel** [reply]
wow, this is poetry :) I did not even try this, I thought we can only load IL code via Load... methods. 
thanks a lot !

if I pass the pdb, it will work as well ? I mean if I publish the library 

```
dotnet publish -r browser-wasm -c Debug
```

I will get a pdb, but will this be of any use for wasm ?

**Reply by maraf** [reply]
Yes, pdbs work

**Reply by maraf** [reply]
I'm not sure you mixing `.dll` and the webcil format `.wasm` for assemblies will work. Let me know if you hit any problems.

**Reply by bogdancucosel** [reply]
actually I seem to have missed the point. When I saw the .wasm files I thought there was some AOT involved, at least in part. I saw earlier the https://github.com/dotnet/runtime/pull/79416 that implements webcil . 
Now it makes sense. 

If AOT compile the app, I would have no way to dynamically load assemblies that would be AOT as well, since this does not work , right ?

**Reply by maraf** [reply]
MonoAOT operates mixed, it AOT code that is AOT friendly and fallback to interpreter for the rest (so that you can all the dynamic features of 
.NET). All AOTed code is packed with runtime code in dotnet.native.wasm and thus it can't be lazy loaded. Anyway, you can still lazy the managed assembly for the fallback code.

You can skip assembly from AOT compilation by not referencing it and thus it won't be AOTed at all.
